fs/dafsserver/dafsserver.cpp

```diff
@@ -1396,21 +1460,413 @@ class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
+        outMeta.setown(getTypeInfoOutputMetaData(config, "output", false));
+        if (!outMeta)
+            outMeta.set(inMeta);
+
+        maxRowSize = config.getPropInt64("maxRowSize", defaultMaxCsvRowSize) * 1024 * 1024;
+        preserveWhitespace = config.getPropBool("preserveWhitespace");
+
+        if (!config.getProp("csvQuote", csvQuote))
+            fileDesc->queryProperties().getProp("@csvQuote", csvQuote);
+        if (!config.getProp("csvSeparate", csvSeparate))
+            fileDesc->queryProperties().getProp("@csvSeparate", csvSeparate);
+        if (!config.getProp("csvTerminate", csvTerminate))
+            fileDesc->queryProperties().getProp("@csvTerminate", csvTerminate);
+        if (!config.getProp("csvEscape", csvEscape))
+            fileDesc->queryProperties().getProp("@csvEscape", csvEscape);
+
+        headerLines = config.getPropInt64("headerLines"); // really this should be a published attribute too
     }
-    virtual const byte * lookupBlob(unsigned __int64 id) override
+    virtual void init(IPropertyTree &config) override
     {
-        throwUnexpected();
+        PARENT::init(config);
+        /* JCSMORE - this needs resolving properly, currently
+         * diskread filters on actual format, and this is filtering on output(projected) format.
+         * We probably need expected format sent too, and filter on that..
+         */
+        if (config.hasProp("keyFilter"))
+        {
+            const RtlRecord &outRecord = outMeta->queryRecordAccessor(true);
+            filterRow = new RtlDynRow(outRecord);
+            Owned<IPropertyTreeIterator> filterIter = config.getElements("keyFilter");
+            ForEach(*filterIter)
+                filters.addFilter(*record, filterIter->query().queryProp(nullptr));
+        }
+    }
+    virtual StringBuffer &getInfoStr(StringBuffer &out) const override
+    {
+        return out.appendf("csvread[%s]", fileName.get());
+    }
+// IRemoteReadActivity impl.
+    virtual const void *nextRow(MemoryBufferBuilder &outBuilder, size32_t &retSz) override
+    {
+        if (eofSeen)
+        {
+            retSz = 0;
+            return nullptr;
+        }
+        checkOpen();
+        while (!eofSeen && (processed < chooseN))
+        {
+            size32_t lineLength = csvSplitter.splitLine(inputStream, maxRowSize);
+            if (!lineLength)
+                break;
+
+            retSz = translator->translate(outBuilder, *this, *fieldFetcher);
+
+            if (retSz) // could it ever be 0 (skip) ??
+            {
+                const void *ret = outBuilder.getSelf();
+                if (fieldFilterMatch(ret))
+                {
+                    outBuilder.finishRow(retSz);
+                    ++processed;
+                    inputStream->skip(lineLength);
+                    return ret;
+                }
+                else
+                    outBuilder.removeBytes(retSz);
```

Comment:
minor: cleaner might be outBuilder.cancelRow() - which would use the reserved length

Reply:
ok, yes agree that would be cleaner, I'll change.

Reply:
hm, actually.. don't see how that can work.
reserved is already extended (during translate, when it ensures..).
removeBytes is reducing reserved.